velox/functions/prestosql/FindFirst.cpp

```diff
@@ -0,0 +1,257 @@
+          onHit(row, firstMatchingIndex.value());
+        } else {
+          onMiss(row);
+        }
+      });
+    }
+  }
+
+ private:
+  static FOLLY_ALWAYS_INLINE std::optional<std::exception_ptr> getOptionalError(
+      const ErrorVectorPtr& errors,
+      vector_size_t row) {
+    if (errors && row < errors->size() && !errors->isNullAt(row)) {
+      return *std::static_pointer_cast<std::exception_ptr>(
+          errors->valueAt(row));
+    }
+
+    return std::nullopt;
+  }
+
+  // Returns an index of the first matching element or std::nullopt if no
+  // element matches or there was an error evaluating the predicate.
+  std::optional<vector_size_t> findFirstMatch(
+      exec::EvalCtx& context,
+      vector_size_t arrayRow,
+      vector_size_t offset,
+      vector_size_t size,
+      const ErrorVectorPtr& elementErrors,
+      const exec::LocalDecodedVector& matchDecoder) const {
+    for (auto i = 0; i < size; ++i) {
+      auto index = offset + i;
+      if (auto error = getOptionalError(elementErrors, index)) {
+        // Report first error to match Presto's implementation.
+        context.setError(arrayRow, error.value());
+        return std::nullopt;
+      }
+
+      if (!matchDecoder->isNullAt(index) &&
+          matchDecoder->valueAt<bool>(index)) {
+        return index;
+      }
+    }
+
+    return std::nullopt;
+  }
+};
+
+class FindFirstFunction : public FindFirstFunctionBase {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& /*outputType*/,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto flatArray = prepareInputArray(args[0], rows, context);
+
+    // Collect indices of the first matching elements or NULLs if no match or
+    // error.
+    BufferPtr resultIndices = allocateIndices(rows.end(), context.pool());
+    auto* rawResultIndices = resultIndices->asMutable<vector_size_t>();
+
+    BufferPtr resultNulls = nullptr;
+    uint64_t* rawResultNulls = nullptr;
+
+    doApply(
+        rows,
+        flatArray,
+        *args[1]->asUnchecked<FunctionVector>(),
+        context,
+        [&](vector_size_t row, vector_size_t firstMatchingIndex) {
+          if (flatArray->elements()->isNullAt(firstMatchingIndex)) {
+            try {
+              VELOX_USER_FAIL("find_first found NULL as the first match");
+            } catch (const VeloxUserError& exception) {
+              context.setVeloxExceptionError(row, std::current_exception());
+            }
+          } else {
+            rawResultIndices[row] = firstMatchingIndex;
+          }
+        },
+        [&](vector_size_t row) {
+          if (rawResultNulls == nullptr) {
+            resultNulls = allocateNulls(rows.end(), context.pool());
+            rawResultNulls = resultNulls->asMutable<uint64_t>();
+          }
+          bits::setNull(rawResultNulls, row);
+        });
+
+    auto localResult = BaseVector::wrapInDictionary(
```

Comment:
Dictionary encoding is always used here to wrap the results. Is it because the cardinality has a high probability of being very low?

Reply:
I use dictionary wrapping here to avoid copying array elements. A potential optimization would be to check the element type and copy if primitive.